SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityDeployChecker.java

```diff
@@ -370,74 +386,211 @@ private SingularityDeployResult getDeployResult(final SingularityRequest request
+          PendingType.NEXT_DEPLOY_STEP, deploy.isPresent() ? deploy.get().getSkipHealthchecksOnDeploy() : Optional.<Boolean> absent(),
+          pendingDeploy.getDeployMarker().getMessage()));
       }
+      return new SingularityDeployResult(DeployState.WAITING);
+    } else {
+
+      final boolean isDeployOverdue = isDeployOverdue(pendingDeploy, deploy);
+      if (deployActiveTasks.size() >= deployProgress.getTargetActiveInstances()) {
+        if (shouldCheckLbState(pendingDeploy)) {
+          final SingularityLoadBalancerUpdate lbUpdate = lbClient.getState(getLoadBalancerRequestId(pendingDeploy));
+
+          updateLoadBalancerStateForTasks(deployActiveTasks, LoadBalancerRequestType.ADD, lbUpdate);
+
+          DeployState deployState = interpretLoadBalancerState(lbUpdate, pendingDeploy.getCurrentDeployState());
+
+          if (deployState == DeployState.SUCCEEDED) {
+            updatePendingDeploy(pendingDeploy, Optional.of(lbUpdate), DeployState.WAITING);
+            return markStepFinished(pendingDeploy, deploy, otherActiveTasks, request, updatePendingDeployRequest);
+          } else if (deployState == DeployState.WAITING) {
+            updatePendingDeploy(pendingDeploy, Optional.of(lbUpdate), deployState);
+            maybeUpdatePendingRequest(pendingDeploy, deploy, request, updatePendingDeployRequest, Optional.of(lbUpdate));
+            return new SingularityDeployResult(DeployState.WAITING);
+          } else {
+            updatePendingDeploy(pendingDeploy, Optional.of(lbUpdate), deployState);
+            maybeUpdatePendingRequest(pendingDeploy, deploy, request, updatePendingDeployRequest, Optional.of(lbUpdate));
+            return fromLbState(deployState, lbUpdate);
+          }
+        }
+
+        if (isDeployOverdue) {
+          if (request.isLoadBalanced() && shouldCancelLoadBalancer(pendingDeploy)) {
+            return cancelLoadBalancer(pendingDeploy);
+          }
+        }
+
+        if (isWaitingForCurrentLbRequest(pendingDeploy)) {
+          return new SingularityDeployResult(DeployState.WAITING, Optional.of("Waiting on load balancer API"), pendingDeploy.getLastLoadBalancerUpdate(), System.currentTimeMillis());
+        }
 
-      if (isCancelRequestPresent) {
-        LOG.info("Canceling a deploy {} due to cancel request {}", pendingDeploy, cancelRequest.get());
-        return new SingularityDeployResult(DeployState.CANCELED, String.format("Canceled due to request by %s at %s", cancelRequest.get().getUser(), cancelRequest.get().getTimestamp()));
+        final DeployHealth deployHealth = deployHealthHelper.getDeployHealth(request, deploy, deployActiveTasks, true);
+        switch (deployHealth) {
+          case WAITING:
+            maybeUpdatePendingRequest(pendingDeploy, deploy, request, updatePendingDeployRequest);
+            return checkOverdue(deploy, deployActiveTasks, isDeployOverdue);
+          case HEALTHY:
+            if (request.isLoadBalanced()) {
```

Comment:
if you can find ways to reduce the massive IF / ELSE indentation on here, great. 

1) through adding additional method calls
2) to inverting - handle the ! case which simply returns something, and then the next code doesn't have to be indented.